### PR TITLE
fix(sdk): Support per workflow TTL (ttl_seconds_after_finished) with new format of Argo workflow manifest

### PR DESCRIPTION
**Pull request by SebBlin**

### CODE REVIEW

**Review by Arhell — COMMENTED** [review]
/ok-to-test

**Comment by SebBlin** [comment]
Hi @Arhell I'm a newbie and this is my first PR. Am I supposed to do something else to make this PR progress or is it fine for my ? Sorry for this question.

File: sdk/python/kfp/compiler/compiler.py [hunk 1]
```diff
@@ -832,8 +832,8 @@ def _create_pipeline_workflow(self,
 
         # set ttl after workflow finishes
         if pipeline_conf.ttl_seconds_after_finished >= 0:
-            workflow['spec'][
-                'ttlSecondsAfterFinished'] = pipeline_conf.ttl_seconds_after_finished
```
Comment (chensun) [inline sdk/python/kfp/compiler/compiler.py:836]: This seems to be a breaking change, do we know from which Argo version was this change introduced? Consequently, which KFP version has that Argo upgrade? We'll need to document this change in the [release note](https://github.com/kubeflow/pipelines/blob/master/sdk/RELEASE.md).

File: sdk/python/kfp/compiler/compiler.py [hunk 2]
```diff
@@ -832,8 +832,8 @@ def _create_pipeline_workflow(self,
...
+            workflow['spec']['ttlStrategy'] = {}
```
Comment (chensun) [inline sdk/python/kfp/compiler/compiler.py:835]: nit: the following line can be combined into this one as 
`workflow['spec']['ttlStrategy'] = {'secondsAfterCompletion' = pipeline_conf.ttl_seconds_after_finished}`

**Comment by SebBlin** [comment]
Here are details about this breaking change:
- On Argo size, this breaking change occurred with version 3.0, see https://argoproj.github.io/argo-workflows/upgrading/#f31e0c6f9-chore-remove-deprecated-fields-5035
- On Kubeflow side, the migration from Argo 2.x to 3.x occured between version 1.3 and 1.4 see manifests :
  - KF 1.3 : https://github.com/kubeflow/manifests/blob/29e6f1f4d51f8dbe33d4aa1f9a53db1d305051e8/apps/pipeline/upstream/third-party/argo/base/workflow-controller-deployment-patch.yaml#L10
  - KF 1.4 : https://github.com/kubeflow/manifests/blob/ca48b5eb6b77e74c957bf4522e47ae82ed78cc4a/apps/pipeline/upstream/third-party/argo/base/workflow-controller-deployment-patch.yaml#L10

I'm not sure where to put this in the release note; I would put it in "Current Version"/"Bug Fixes and Other Changes" as this is a fix that correct a breaking from Argo; it doesn't introduce a breaking change in KFP. 
Is this correct?

**Comment by chensun** [comment]
> Here are details about this breaking change:
> 
> * On Argo size, this breaking change occurred with version 3.0, see https://argoproj.github.io/argo-workflows/upgrading/#f31e0c6f9-chore-remove-deprecated-fields-5035
> * On Kubeflow side, the migration from Argo 2.x to 3.x occured between version 1.3 and 1.4 see manifests :
>   
>   * KF 1.3 : https://github.com/kubeflow/manifests/blob/29e6f1f4d51f8dbe33d4aa1f9a53db1d305051e8/apps/pipeline/upstream/third-party/argo/base/workflow-controller-deployment-patch.yaml#L10
>   * KF 1.4 : https://github.com/kubeflow/manifests/blob/ca48b5eb6b77e74c957bf4522e47ae82ed78cc4a/apps/pipeline/upstream/third-party/argo/base/workflow-controller-deployment-patch.yaml#L10
> 
> I'm not sure where to put this in the release note; I would put it in "Current Version"/"Bug Fixes and Other Changes" as this is a fix that correct a breaking from Argo; it doesn't introduce a breaking change in KFP. Is this correct?

Thanks for the info. On KFP side, upgrading Argo workflow from 2.x to 3.x was done by https://github.com/kubeflow/pipelines/pull/5922, which in in KFP (backend) 1.7.0-alpha.1 release per [the changelog](https://github.com/kubeflow/pipelines/blob/master/CHANGELOG.md#170-alpha1-2021-06-28). That means this SDK change would make the next KFP SDK work with KFP backend 1.7+, but incompatible with earlier version. In that sense, it's  a breaking change for KFP pre-1.7 versions. Does that make sense?

**Comment by SebBlin** [comment]
>
> That means this SDK change would make the next KFP SDK work with KFP backend 1.7+, but incompatible with earlier version. In that sense, it's a breaking change for KFP pre-1.7 versions. Does that make sense?

Indeed, you are completely right. I updated the changelog accordingly

**Comment by chensun** [comment]
/lgtm
/approve

Thanks!

**Pull request merged**